### PR TITLE
Sample weight checks removal from Kmeans predict

### PR DESCRIPTION
**Pull request by md-shafiul-alam**
### Description

check_sample_weights or np_allclose is pretty expensive and have performance impact even if the branch is not taken. Since sklearn dropped support for sample weights in KMeans Predict from 1.5, we can ignore all these expensive checks. Fit continues to support sample_weight though.

_List issue number(s) if exist(s): #6 (for example)_

---

Checklist to comply with before moving PR from draft:

**PR completeness and readability**

- [ ] I have reviewed my changes thoroughly before submitting this pull request.
- [ ] I have commented my code, particularly in hard-to-understand areas.
- [ ] I have updated the documentation to reflect the changes or created a separate PR with update and provided its number in the description, if necessary.
- [ ] Git commit message contains an appropriate signed-off-by string _(see [CONTRIBUTING.md](https://github.com/intel/scikit-learn-intelex/blob/main/CONTRIBUTING.md#pull-requests) for details)_.
- [ ] I have added a respective label(s) to PR if I have a permission for that.  
- [ ] I have resolved any merge conflicts that might occur with the base branch.

**Testing**

- [ ] The unit tests pass successfully.
- [ ] I have run it locally and tested the changes extensively.

**Performance**

- [ ] I have measured performance for affected algorithms using [scikit-learn_bench](https://github.com/IntelPython/scikit-learn_bench) and provided at least summary table with measured data, if performance change is expected.
- [ ] I have provided justification why performance has changed or why changes are not expected.

### CODE REVIEW

**Comment by icfaust** [comment]
I assume its an extension of this: https://github.com/intel/scikit-learn-intelex/pull/1770#discussion_r1652169125 

How is this code being run unless sample_weights is not None, are we just removing a check just for performance purposes?

**Comment by md-shafiul-alam** [comment]
with this change the **predict** will dispatch to sklearn for sample_weight != None. We will lose coverage for np_ones for sklearn <1.5. Still working to figure out if this improves perf.

File: sklearnex/cluster/k_means.py [hunk 1]
```diff
@@ -213,8 +204,8 @@ def _onedal_predict_supported(self, method_name, X, sample_weight=None):
                         "Supported data formats: Dense, CSR (oneDAL version >= 2024.7.0).",
                     ),
                     (
-                        _acceptable_sample_weights,
-                        "oneDAL doesn't support sample_weight, None or ones are acceptable",
+                        sample_weight is None,
+                        "oneDAL doesn't support sample_weight.",
```
Comment (samir-nasibli) [inline sklearnex/cluster/k_means.py:208]: As the same question on the above. Did something changed on onedal side to update this message? Was this this a bug?

File: sklearnex/cluster/k_means.py [hunk 2]
```diff
@@ -109,14 +109,11 @@ def _onedal_fit_supported(self, method_name, X, y=None, sample_weight=None):
...
             ) or not issparse(X)
 
-            _acceptable_sample_weights = True
-            if sample_weight is not None or not isinstance(sample_weight, numbers.Number):
-                sample_weight = _check_sample_weight(
-                    sample_weight, X, dtype=X.dtype if hasattr(X, "dtype") else None
-                )
-                _acceptable_sample_weights = np.allclose(
-                    sample_weight, np.ones_like(sample_weight)
-                )
```
Comment (samir-nasibli) [inline sklearnex/cluster/k_means.py:119]: Instead of removing it I prefer to branch the logic for different versions of sklearn.
For the skl 1.5 we can skip this. Did you tried this?

File: sklearnex/cluster/k_means.py [hunk 3]
```diff
@@ -127,7 +124,7 @@ def _onedal_fit_supported(self, method_name, X, y=None, sample_weight=None):
                     (correct_count, "n_clusters is smaller than number of samples"),
                     (
                         _acceptable_sample_weights,
-                        "oneDAL doesn't support sample_weight, either None or ones are acceptable",
+                        "oneDAL doesn't support sample_weight, None or equal weights are accepted.",
```
[the remark this answers is not in the document]
Comment (md-shafiul-alam) [reply sklearnex/cluster/k_means.py:127]: Basically the message includes the constant value option for sample_weight. constant sample_weight is transformed to an array of equal values.

File: sklearnex/cluster/k_means.py [hunk 4]
```diff
@@ -161,6 +158,9 @@ def _onedal_fit(self, X, _, sample_weight, queue=None):
                 X,
                 accept_sparse="csr",
                 dtype=[np.float64, np.float32],
+                order="C",
+                copy=self.copy_x,
+                accept_large_sparse=False,
```
[the remark this answers is not in the document]
Comment (md-shafiul-alam) [reply sklearnex/cluster/k_means.py:163]: This is to align the check with daal4py and sklearn

File: sklearnex/cluster/k_means.py (hunk 2 above)
Comment (md-shafiul-alam) [reply sklearnex/cluster/k_means.py:119]: Interesting finding, branching doesn't help. The interpreter still spends time even if the branch is not taken. The behavior is same in daal4py.

File: sklearnex/cluster/k_means.py (hunk 1 above)
Comment (md-shafiul-alam) [reply sklearnex/cluster/k_means.py:208]: Nothing changed in oneDAL, the checks were removed because they are causing significant perf regression. The message is more appropriate.

File: sklearnex/cluster/k_means.py (hunk 2 above)
Comment (md-shafiul-alam) [reply sklearnex/cluster/k_means.py]: Updated with version check branching. Note that this is slightly different from what daal4py has. Anyway should work fine.

Comment (icfaust) [reply sklearnex/cluster/k_means.py]: > Interesting finding, branching doesn't help. The interpreter still spends time even if the branch is not taken. The proposed change aligns with daal4py.

i find that really unusual/ not sure how thats even possible. have you done any trace or profiling to show this?

Comment (md-shafiul-alam) [reply sklearnex/cluster/k_means.py]: My assumption was wrong, the "or" was the problem.

Comment (icfaust) [reply sklearnex/cluster/k_means.py]: ahhhhh i see now. thanks!

Comment (samir-nasibli) [reply sklearnex/cluster/k_means.py]: > check branching. Note that

Functionally, it should be aligned first with the stock sklearn. This is more prior. Thank you!

File: sklearnex/cluster/k_means.py [hunk 5]
```diff
@@ -110,13 +110,13 @@ def _onedal_fit_supported(self, method_name, X, y=None, sample_weight=None):
...
+            if (sample_weight is not None) and (
+                not isinstance(sample_weight, numbers.Number)
+            ):
                 sample_weight = _check_sample_weight(
                     sample_weight, X, dtype=X.dtype if hasattr(X, "dtype") else None
                 )
-                _acceptable_sample_weights = np.allclose(
-                    sample_weight, np.ones_like(sample_weight)
-                )
+                _acceptable_sample_weights = np.all(sample_weight == 1)
```
Comment (samir-nasibli) [inline sklearnex/cluster/k_means.py:119]: This is not applicable for constant values that differ from 1?
```suggestion
                _acceptable_sample_weights = np.all(sample_weight == sample_weight[0])
```

**Review by samir-nasibli — COMMENTED** [review]
Expecting the latest perf validation. Generally looks good to me. Thank you!

**Review by samir-nasibli — APPROVED** [review]

**Review by icfaust — APPROVED** [review]
Check for approval with @syakov-intel @napetrov before merge

**Review by syakov-intel — APPROVED** [review]
approved to merge
